synapse/lib/ast.py

```
@@ -2758,3 +2849,104 @@ def prepare(self):
+        if self.kids:
+            valu = await self.kids[0].compute(runt)
+
+        raise StormReturn(valu)
+
+class FuncArgs(AstNode):
+
+    def value(self):
+        return [k.value() for k in self.kids]
+
+class Function(AstNode):
+    '''
+    ( name, args, body )
+
+    // use args/kwargs syntax
+    function bar(x, v=$(30)) {
+    }
+
+    # we auto-detect the behavior of the target function
+
+    # return a value
+    function bar(x, y) { return $(x + y) }
+
+    # incrementally yield values
+    function bar(x, y) { yield $x yield $y }
+
+    # a function that produces nodes
+    function bar(x, y) { [ baz:faz=(x, y) ] }
+
+    $foo = $bar(10, v=20)
+    '''
+    async def run(self, runt, genr):
+
+        self.hasretn = self.hasAstClass(Return)
+
+        async def realfunc(*args, **kwargs):
+            return await self.callfunc(runt, args, kwargs)
+
+        name = self.kids[0].value()
+        runt.setVar(name, realfunc)
+
+        async for node, path in genr:
+            path.setVar(name, realfunc)
+            yield node, path
+
+    def getRuntVars(self, runt):
+        yield self.kids[0].value()
+
+    async def callfunc(self, runt, args, kwargs):
+        '''
+        Execute a function call using the given runtime.
+
+        This function may return a value / generator / async generator
+        '''
+        argdefs = self.kids[1].value()
+        if len(args) != len(argdefs):
+            raise Exception('BAD CALL ARG COUNT FIXME')
+
+        argvars = {}
+        for i, name in enumerate(argdefs):
+            argvars[name] = args[i]
+
+        opts = {'vars': argvars}
+        funcrunt = await runt.getScopeRuntime(self.kids[2], opts=opts)
+
+        genr = agen()
+
+        if self.hasretn:
+
+            try:
+
+                async for item in self.kids[2].run(funcrunt, agen()):
+                    pass
+
+            except StormReturn as e:
+                return e.item
+
+            return None
```

should this raise instead of return?